src/shared/tpm2-util.c

```diff
@@ -1395,6 +1395,84 @@ static void hash_pin(const char *pin, size_t len, TPM2B_AUTH *auth) {
         sha256_finish_ctx(&hash, auth->buffer);
 }
 
+/* Hash data into the digest.
+ *
+ * If 'extend' is true, the hashing operation starts with the existing digest hash (and the digest is
+ * required to have a hash and its size must be correct). If 'extend' is false, the digest size is
+ * initialized to the correct size for 'alg' and the hashing operation does not include any existing digest
+ * hash. If 'extend' is false and no data is provided, the digest is initialized to a zero digest.
+ *
+ * On success, the digest hash will be updated with the hashing operation result and the digest size will be
+ * correct for 'alg'.
+ *
+ * This currently only provides SHA256, so 'alg' must be TPM2_ALG_SHA256. */
+int tpm2_digest_hash_buffers(
+                TPMI_ALG_HASH alg,
+                TPM2B_DIGEST *digest,
+                const uint8_t *data[],
+                const size_t len[],
+                size_t count,
+                bool extend) {
+
+        struct sha256_ctx ctx;
+
+        assert(digest);
+        assert(data != NULL || count == 0);
+        assert(len != NULL || count == 0);
+
+        if (alg != TPM2_ALG_SHA256)
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "Hash alg not supported: 0x%x", alg);
```

Comment:
typically we say that "library-like" functions shouldn't log on their own beyond LOG_DEBUG, but let the caller do that. It's not always clear where to draw the line between "library-like" and "main-function-like", but this one seems over the line, no? dunno?